Cache student scores at student level.

dandrabik opened the pull request:
## WHAT
Edit: Also optimize queries to run one group query, instead of looping through many queries.
Adjust student score caching to be at the student level (not student/activity/classroom_unit level). 
## WHY
This controller action is still slow. We're spending a lot of time in Redis per controller action and making an average of 30 calls. This should reduce that time drastically and make subsequent calls faster. The downside is this cache will be flushed more often (when ever a new activity is finished, the whole cache for a student is flushed, vs. just a section of it), but I think it will be a net gain.
## HOW
Move the caching outside of the loop. Do one main query, put the results in a hash and look them up.
### Screenshots
<img width="1291" alt="Screenshot 2024-03-20 at 4 47 28 PM" src="https://github.com/empirical-org/Empirical-Core/assets/1304933/cc699d9a-2fc7-4175-a78e-aac9f380331a">


### Notion Card Links
(Please provide links to any relevant Notion card(s) relevant to this PR.)

### What have you done to QA this feature?
I QA'd this feature on my staging and will ask Emilia to do a quick QA as well.

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | YES
Self-Review: Have you done an initial self-review of the code below on Github? | Yes

dandrabik committed:
Cache student scores at student level.

dandrabik committed:
Whitespace.

dandrabik commented:
@emilia-friedberg If you get a chance tomorrow morning, could you do a quick QA on staging (it's deployed to my env).

emilia-friedberg reviewed (changes requested):
Hey Dan -- I'm happy to QA this, but I think we need to be caching by at least the classroom id in addition to the student. Right now, this call is made for every classroom (that's how the "scores" query that this is built on is scoped). Otherwise, changing classrooms would result in the wrong data showing. We can either update the frontend to pass back the classroom id as one of the params (let me know if you'd like me to work on that), or find the first unit activity's classroom unit and grab the classroom id from there.

dandrabik commented:
> Hey Dan -- I'm happy to QA this, but I think we need to be caching by at least the classroom id in addition to the student. Right now, this call is made for every classroom (that's how the "scores" query that this is built on is scoped). Otherwise, changing classrooms would result in the wrong data showing. We can either update the frontend to pass back the classroom id as one of the params (let me know if you'd like me to work on that), or find the first unit activity's classroom unit and grab the classroom id from there.

Ah, ok, good catch. Let me take a pass at this. It's not super urgent since the page is stable now, but will look at this later today. Thanks for the review.

dandrabik committed:
Reduce queries, cache by classroom_id.

dandrabik committed:
Refactor for complexity.

dandrabik committed:
Add nil protection.

dandrabik committed:
Eager load two more relations.

dandrabik commented:
> > Hey Dan -- I'm happy to QA this, but I think we need to be caching by at least the classroom id in addition to the student. Right now, this call is made for every classroom (that's how the "scores" query that this is built on is scoped). Otherwise, changing classrooms would result in the wrong data showing. We can either update the frontend to pass back the classroom id as one of the params (let me know if you'd like me to work on that), or find the first unit activity's classroom unit and grab the classroom id from there.
> 
> Ah, ok, good catch. Let me take a pass at this. It's not super urgent since the page is stable now, but will look at this later today. Thanks for the review.

@emilia-friedberg I've adjusted the caching on this page and also added some query optimization (one group query put into a hash that the rest pull from). This PR is ready for re-review. I've also done a basic QA on my staging env, but would love a second set of eyes since I'm not too familiar with this feature.

dandrabik committed:
Lint refactor.

emilia-friedberg committed:
make sure we're passing the classroomid to fetchexactscoresdata

emilia-friedberg commented on services/QuillLMS/app/controllers/profiles_controller.rb:
```diff
@@ -87,34 +87,44 @@ def staff
-        'classroom_unit_id' => classroom_unit_id,
-        'ua_id' => ua_id
-      }
+  ACTIVITY_ID = 'activity_id'
+  CLASSROOM_UNIT_ID = 'classroom_unit_id'
+  UA_ID = 'ua_id'
```
I feel like we usually group constants at the top of a file. I'm not sure it really matters either way because I'm normally cmd+f-ing anyway, but we might want to move these for consistency.

dandrabik replied:
That's fair. I've starting putting constants close to their use, but I think that breaks from the pattern we use everywhere else, so I'll move them to the top.

emilia-friedberg commented on services/QuillLMS/app/controllers/profiles_controller.rb:
```diff
@@ -87,34 +87,44 @@ def staff
+  private def student_score_cache_key = User.student_scores_cache_key(current_user.id, params[:classroom_id])
+
+  private def exact_scores_data_all(user, data, classroom_id)
+    activity_sessions_grouped = student_activity_sessions_grouped(user.id, data)
```
Yay for the efficiency gain here! Good call.

emilia-friedberg reviewed (approved):
Deployed the most recent version to my staging environment, QAed, and all looks well. One tiny note about constant placement but otherwise LGTM!

dandrabik committed:
Move constants to top of file.
